### ISSUE TITLE
middleware matcher should support template literals

### ISSUE THREAD

**Issue by stefanprobst** [vercel/next.js#56398]
### Link to the code that reproduces this issue

https://github.com/stefanprobst/issue-next-middleware-matcher

### To Reproduce

1. clone https://github.com/stefanprobst/issue-next-middleware-matcher
2. `pnpm install && pnpm run dev`
3. see error message in console: 'Next.js can't recognize the exported `config` field in route "/middleware": Unsupported template literal with expressions at "config.matcher[1]".'

### Current vs. Expected behavior

i would want to be able to do this:

```ts
const locales = ["de", "en"] as const;

export const config = {
  matcher: ["/", `/(${locales.join("|")})/:path*`],
};
```

### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

```bash
Operating System:
  Platform: linux
  Arch: x64
  Version: #26~22.04.1-Ubuntu SMP PREEMPT_DYNAMIC Thu Jul 13 16:27:29 UTC 2
Binaries:
  Node: 18.18.0
  npm: 9.8.1
  Yarn: 1.22.19
  pnpm: 8.8.0
Relevant Packages:
  next: 13.5.4
  eslint-config-next: N/A
  react: 18.2.0
  react-dom: 18.2.0
  typescript: 5.2.2
Next.js Config:
  output: N/A
```

### Which area(s) are affected? (Select all that apply)

Middleware / Edge (API routes, runtime)

**Comment by hydRAnger** [comment]
same here, as I'd like to organize some complex matcher rules with js.
then I faced this issue.

**Comment by arduano** [comment]
A simpler reproduction:
```ts
const base = '/api';
export const config = {
  matcher: `${base}/:path*`,
};
```

Substituting `${base}` with the string directly works.

**Comment by SalahAdDin** [comment]
Same issue here:
```
import createMiddleware from "next-intl/middleware";

import { DEFAULT_LOCALE, LOCALES, LOCALES_TO_MATCH } from "./common/constants";

export default createMiddleware({
  locales: LOCALES,

  defaultLocale: DEFAULT_LOCALE,
});

export const config = {
  matcher: [
    "/",
    `/(${LOCALES_TO_MATCH})/:path*`,
    "/((?!_next|_vercel|.*\\..*).*)",
  ],
};
```

Yeah, same issue here, required for most of the current internationalization packages.

**Comment by gmbrown** [comment]
fwiw, these docs do state pretty clearly that you cannot do this
https://nextjs.org/docs/messages/invalid-page-config

That being said, I'm not sure why and I ended up here because I want to be able to

**Comment by SalahAdDin** [comment]
> fwiw, these docs do state pretty clearly that you cannot do this https://nextjs.org/docs/messages/invalid-page-config
> 
> That being said, I'm not sure why and I ended up here because I want to be able to

We need a workaround on it, since, just for instance, our project will contain a minimum of 30 locales, and we cannot add them manually.

**Comment by SalahAdDin** [comment]
> A simpler reproduction:
> 
> ```ts
> const base = '/api';
> export const config = {
>   matcher: `${base}/:path*`,
> };
> ```
> 
> Substituting `${base}` with the string directly works.

You say this solution works, right?

**Comment by aryankarim** [comment]
Can somebody remind us when there is support for string matutation? As of today, there is no support.

**Comment by mwx0621** [comment]
This problem also occurs to me when trying to migrate from i18next to next-intl. In my case, different locales are supported in different deployment. Matchers need to be set from config file.
I solved this by manually matching pathname in middleware function. 
```typescript
import createMiddleware from "next-intl/middleware";
import { LOCALES } from "./i18n";
import { NextRequest, NextResponse } from "next/server";

const reg = new RegExp(`^/$|^/(${LOCALES.join("|")})`);

const intlMiddleware = createMiddleware({
  locales: LOCALES,
  defaultLocale: "en",
});

const middleware = (req: NextRequest) => {
  const { pathname } = new URL(req.url);
  if (pathname.match(reg)) {
    return intlMiddleware(req);
  } else return NextResponse.next();
};
export default middleware;

// Which I wish it could be 
// export const config = {
//   matcher: ["/", `/(${LOCALES.join("|"))/:path*`],
// };
```
This works for me.

**Comment by SalahAdDin** [comment]
> This problem also occurs to me when trying to migrate from i18next to next-intl. In my case, different locales are supported in different deployment. Matchers need to be set from config file. I solved this by manually matching pathname in middleware function.
> 
> ```ts
> import createMiddleware from "next-intl/middleware";
> import { LOCALES } from "./i18n";
> import { NextRequest, NextResponse } from "next/server";
> 
> const reg = new RegExp(`^/$|^/(${LOCALES.join("|")})`);
> 
> const intlMiddleware = createMiddleware({
>   locales: LOCALES,
>   defaultLocale: "en",
> });
> 
> const middleware = (req: NextRequest) => {
>   const { pathname } = new URL(req.url);
>   if (pathname.match(reg)) {
>     return intlMiddleware(req);
>   } else return NextResponse.next();
> };
> export default middleware;
> 
> // Which I wish it could be 
> // export const config = {
> //   matcher: ["/", `/(${LOCALES.join("|"))/:path*`],
> // };
> ```
> 
> This works for me.

Where is the `export const config = {}` there?

**Comment by mwx0621** [comment]
> > This problem also occurs to me when trying to migrate from i18next to next-intl. In my case, different locales are supported in different deployment. Matchers need to be set from config file. I solved this by manually matching pathname in middleware function.
> > ```ts
> > import createMiddleware from "next-intl/middleware";
> > import { LOCALES } from "./i18n";
> > import { NextRequest, NextResponse } from "next/server";
> > 
> > const reg = new RegExp(`^/$|^/(${LOCALES.join("|")})`);
> > 
> > const intlMiddleware = createMiddleware({
> >   locales: LOCALES,
> >   defaultLocale: "en",
> > });
> > 
> > const middleware = (req: NextRequest) => {
> >   const { pathname } = new URL(req.url);
> >   if (pathname.match(reg)) {
> >     return intlMiddleware(req);
> >   } else return NextResponse.next();
> > };
> > export default middleware;
> > 
> > // Which I wish it could be 
> > // export const config = {
> > //   matcher: ["/", `/(${LOCALES.join("|"))/:path*`],
> > // };
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > This works for me.
> 
> Where is the `export const config = {}` there?

I didn't export a config here, I think that next will fallback to a default config which will match all pathname.
Or you can try to leave an config with no matcher declared here.
```typescript
export const config = {
    // configurations except matcher
}
```

**Comment by dgordon-p1x** [comment]
> This problem also occurs to me when trying to migrate from i18next to next-intl. In my case, different locales are supported in different deployment. Matchers need to be set from config file. I solved this by manually matching pathname in middleware function.
> 
> ```ts
> import createMiddleware from "next-intl/middleware";
> import { LOCALES } from "./i18n";
> import { NextRequest, NextResponse } from "next/server";
> 
> const reg = new RegExp(`^/$|^/(${LOCALES.join("|")})`);
> 
> const intlMiddleware = createMiddleware({
>   locales: LOCALES,
>   defaultLocale: "en",
> });
> 
> const middleware = (req: NextRequest) => {
>   const { pathname } = new URL(req.url);
>   if (pathname.match(reg)) {
>     return intlMiddleware(req);
>   } else return NextResponse.next();
> };
> export default middleware;
> 
> // Which I wish it could be 
> // export const config = {
> //   matcher: ["/", `/(${LOCALES.join("|"))/:path*`],
> // };
> ```
> 
> This works for me.

Brilliant! Thanks @mwx0621 that did it! Great idea to manually invoke the `intlMiddleware` when need be 🧠

I made some adjustments to remove unnecessary RegEx matching and url creation.


```tsx
import createMiddleware from "next-intl/middleware";
import { routing } from "i18n/routing";
import { NextResponse, type NextRequest } from "next/server";

const intlMiddleware = createMiddleware(routing);

const middleware = (req: NextRequest) => {
  const { pathname } = req.nextUrl;
  if (
    pathname === "/" ||
    routing.locales.some((locale) => pathname.startsWith(`/${locale}`))
  )
    return intlMiddleware(req);
  return NextResponse.next();
};

export default middleware;
```
Hope this helps someone!

**Comment by theoludwig** [comment]
What is the root of the issue? As middleware matchers are declared in JS/TS, why JS/TS doesn't work?

Are the matchers behavior the same with both edge and Node.js runtime?

**Comment by icyJoseph** [comment]
Let's try to retake the conversation here. 

The `config` export in middleware/proxy has to be **statically analyzable**.  

This means Next.js reads it at build/dev time by parsing the file’s AST, without executing the module.

That config is **not** consumed by the runtime middleware function. Instead, it’s used earlier by the build pipeline to generate routing manifests (like `matcher`) and decide how the middleware bundle should be produced and deployed. 

Because of that, Next needs to extract the config directly from syntax, not from evaluated JavaScript.

A declaration like this works:

```ts
export const config = {
  matcher: ['/about'],
}
````

because the compiler can find an `export const config = /* literally the configuration */` node and read its value directly from the AST.

But patterns like these don’t work:

```ts
const config2 = { matcher: ['/about'] }
export const config = config2

const config = { matcher: ['/about'] }
export { config }
```

because now the extractor would need to resolve identifiers or follow re-exports. That requires symbol resolution (and potentially executing code), which the middleware/proxy config parser intentionally avoids. 

If it can’t match the exact export shape, it behaves as if no config was provided (leading to a build error I believe). In other words, a smarter parser could handle these two examples. But once you open that door, users will reasonably expect lots of JS to work, and the maintenance/correctness surface area explodes, as we can see in this thread.

This strict approach to keep things simple, fast and predictable. 

The work around to do the check within the middleware/proxy function, ought to work, but it must be paired with a matcher still, to prevent firing the middleware unnecessarily for:

- static assets
- other route branches

In other words, you'd allow a path for locales, but not define exactly all locales in the config matcher, but do the last pass check within the function.

**Comment by ChihebBENCHEIKH1** [comment]
I've opened a PR to address this: #93150

The fix adds support for template literal expressions in the static config extractor, including module-level `const` identifier resolution and `Array.join()` evaluation. This enables the common i18n pattern:

```typescript
const locales = ['en', 'de', 'fr'] as const
export const config = {
  matcher: ['/', `/(${locales.join('|')})/:path*`],
}
```
